### PR TITLE
facilitator integration tester & terraform changes

### PR DESCRIPTION
**Pull request by aaomidi**
#226 

Changes in this PR:
- Removal of the sample-maker
- deploy-tool has to now be ran once after `make apply-bootstrap`, and once after `make apply` in terraform
- cluster initial_node_count, min_node_count, and max_node_count are now configurable in tfvars
- facilitator can now read DataShareSpecificManifests and bootstrap accordingly
- updates the manifest-updater's (still unused) manifest signatures
- `make apply-bootstrap` now creates the GKE instance as well
- Updates the terraform in CI to 0.14.4

The main story of this PR is the apple-style integration-tester. This PR creates a semi real-world scenario of how an ingestor works and allows us to do end to end testing of our deployments and changes.

### CODE REVIEW

File: terraform/main.tf
Comment (aaomidi) [inline terraform/main.tf:368]: I found these variables made reasoning about the environment a lot easier.

File: terraform/modules/fake_server_resources/fake_server_resources.tf
Comment (tgeoghegan) [inline terraform/modules/fake_server_resources/fake_server_resources.tf:169]: The resources `aws_iam_role_policy.bucket_role_policy` and `aws_iam_role.tester_role` duplicate a lot of code (down to the comments) from `data_share_processor.tf`. Perhaps it could be factored out into a small module?

**Review by tgeoghegan — CHANGES_REQUESTED** [review]
So this approach creates a cronjob that creates `integration-tester` jobs which then each create a further `facilitator` job, which seems a little complicated. I hate to come in with this observation after you got all this working, but why didn't we just put all the necessary logic into `facilitator` itself, and save ourselves the extra job management overhead, not to mention a bunch of code?

File: workflow-manager/kubernetes/kubernetes.go
Comment (aaomidi) [reply workflow-manager/kubernetes/kubernetes.go:97]: Go limitation, it doesn't do recursive replaces sadly: https://github.com/golang/go/issues/39358 I think this describes it.

File: deploy-tool/main.go
Comment (aaomidi) [reply deploy-tool/main.go:332]: Sounds good, I thought since this is only impacting the test environment it's not a big deal if we "reset" essentially.

**Comment by aaomidi** [comment]
> but why didn't we just put all the necessary logic into facilitator itself, and save ourselves the extra job management overhead, not to mention a bunch of code?

Few reasons for this:
- I was more comfortable writing in Go than Rust
- Lack of first party kubernetes API in Rust (I don't 100% need this, but it's useful to have to find the correct batch-signing-key)

I don't think I particularly want to move this into Rust, especially since the details of the implementation were somewhat agreed to in the original design document. I think we can look into moving the google style ingestion server into facilitator and then adding support for the apple style ingestion server into the facilitator as well.

File: deploy-tool/main.go
Comment (aaomidi) [reply deploy-tool/main.go:68]: Good idea yeah

File: deploy-tool/main.go
Comment (aaomidi) [reply deploy-tool/main.go]: I did the same thing we do for the DSP manifests, just check if the manifest exists or not. Will add a bypass flag in a future update.

File: integration-tester/tester/tester.go
Comment (aaomidi) [reply integration-tester/tester/tester.go:52]: I did this but I don't think it matters much because for small micro-services (or micro-jobs? not sure what to call it) like this, if this fails, then we're going to quit anyway.

File: integration-tester/tester/tester.go
Comment (aaomidi) [reply integration-tester/tester/tester.go:238]: Ah because this used to use t. After removing it I forgot to change the signature.

File: terraform/modules/fake_server_resources/fake_server_resources.tf
Comment (aaomidi) [reply terraform/modules/fake_server_resources/fake_server_resources.tf:200]: No, this has more permissions to the jobs resource, and lives in a different namespace

File: deploy-tool/main.go
Comment (aaomidi) [reply deploy-tool/main.go:131]: Hmm that would work. I want to eventually stop making the secrets in terraform though since that's going to have to go away when the deploy-operator goes away. For example I'm not making the secret for the test ingestor in terraform.

I think if our goal is to move secret creation out of terraform we could potentially have this as an interim step?

File: manifest-updater/manifest/types.go
Comment (aaomidi) [reply manifest-updater/manifest/types.go:48]: Ah good to know!

File: terraform/modules/fake_server_resources/fake_server_resources.tf
Comment (tgeoghegan) [reply terraform/modules/fake_server_resources/fake_server_resources.tf:200]: Then why does it reference the `integration-tester`?

Comment (aaomidi) [reply terraform/modules/fake_server_resources/fake_server_resources.tf:200]: Oh that's what you mean. Forgot to clean that part of the code up. Will revet the role in kubernetes.tf to what it used to be.

**Comment by tgeoghegan** [comment]
I'm afraid I'm just not comfortable with the Go cronjob wrapper.
- We learned very painfully that cronjobs that spawn other jobs are an antipattern in Kubernetes, and while this instance won't be as bad as the old workflow-manager design, I don't see why we would choose to repeat that mistake when we could make this a simple Kubernetes deployment.
- We should avoid making architectural decisions based on an individual contributor's comfort with this-or-that language or technology. We're not under the gun to get this out as we have been thus far on this project so we should take the time to do it right. I think [rpki](https://docs.rs/rpki/0.10.0/rpki/csr/struct.Csr.html) should let us parse PKCS#10 CSRs, and there are some basic Rust Kubernetes clients out there (the Kubernetes documentation lists [this one](https://github.com/clux/kube-rs) and [this one](https://github.com/ynqa/kubernetes-rust). In the worst case, implementing the single API we need to list secrets should be no harder than the minimal GCS and PubSub  client I wrote. If you're not comfortable taking on the Rust work, I'm happy to take on the needed facilitator changes.

**Comment by aaomidi** [comment]
> If you're not comfortable taking on the Rust work, I'm happy to take on the needed facilitator changes.

Sure I think that sounds fine then. Do you want me to make another PR with just the terraform/deploy-tool changes or delete those changes from this PR?

**Comment by tgeoghegan** [comment]
The majority of the work in this PR is very valuable -- all the work you did to set up the tester namespace, to construct manifests, to wire up Kubernetes secrets and so on. And your Go implementation of the `integration-tester` at least proves that all the necessary parameters, values, etc. have been plumbed into the right places, regardless of the specific implementation of the tester itself. So I think we could merge it as is (though I need to do another review pass on all the Terraform stuff), and then later this week (time allowing) I will replace the `integration-tester` cronjob with a deployment that runs a new `facilitator generate-ingestion-sample-worker`.

File: deploy-tool/main.go
Comment (tgeoghegan) [reply deploy-tool/main.go:131]: I believe moving secret creation out of Terraform means doing it in `deploy-operator`/`manifest-updater`, and that change would also obsolete `deploy-tool`. So adding this code doesn't actually help us make progress toward the automated secret generation/rotation world (because this new code will get deleted along with the rest of `deploy-tool`, and in the interim, I think it's more complicated than applying the label from Terraform, which I think would probably be a one line change compared to plumbing the value of `ingestorName` all the way from TF to here.

**Comment by aaomidi** [comment]
Hopefully for the locality-specific ingestor I'll be able to build it up in rust. Hoping the ureq v2 work makes me more comfortable with the Rust codebase.

**Review by tgeoghegan — CHANGES_REQUESTED** [review]

File: facilitator/src/bin/facilitator.rs
Comment (aaomidi) [reply facilitator/src/bin/facilitator.rs]: This code path is outdated and the current code doesn't require kubernetes for making samples

File: facilitator/src/bin/facilitator.rs
Comment (aaomidi) [reply facilitator/src/bin/facilitator.rs]: With context I wouldn't have access to `e`, right?

File: facilitator/src/bin/facilitator.rs
Comment (tgeoghegan) [inline facilitator/src/bin/facilitator.rs:986]: Does this mean I can't run `facilitator generate-ingestion-sample` on my laptop anymore? I'd still like to be able to generate sample data without running this in Kubernetes.

File: facilitator/src/bin/facilitator.rs
Comment (tgeoghegan) [inline facilitator/src/bin/facilitator.rs:1597]: We shouldn't log this, even at debug level, to avoid the risk of accidentally disclosing private keys into cloud log storage.

File: facilitator/src/intake.rs
Comment (tgeoghegan) [inline facilitator/src/intake.rs:116]:
```suggestion
                debug!("Public key for server is: {:?}", PublicKey::from(k));
```

File: facilitator/src/intake.rs
Comment (tgeoghegan) [inline facilitator/src/intake.rs:121]:
```suggestion
        debug!("We have {} servers.", &servers.len());
```

File: facilitator/src/kubernetes.rs
Comment (tgeoghegan) [inline facilitator/src/kubernetes.rs:27]: I think this is fine, but I think the more idiomatic Rusty thing would be to implement [`std::convert::From<Secret>`](https://doc.rust-lang.org/std/convert/trait.From.html) on `Time`.

File: facilitator/src/kubernetes.rs
Comment (tgeoghegan) [inline facilitator/src/kubernetes.rs:14]: This should be called something like `KubernetesClient` to make it clear we're not implementing a Kubernetes API server here.

File: facilitator/src/kubernetes.rs
Comment (tgeoghegan) [inline facilitator/src/kubernetes.rs:79]: This should be equivalent:
```
stuff.await.context(format!("listing secrets failed with {} as label_selector", label_selector))?;
```
Up to you which you like better.

File: facilitator/src/sample.rs
Comment (tgeoghegan) [inline facilitator/src/sample.rs:126]: Was this necessary? `.context()` should preserve the underlying error, unless I have missed something.

**Review by tgeoghegan — CHANGES_REQUESTED** [review]

File: facilitator/src/bin/facilitator.rs
Comment (aaomidi) [reply facilitator/src/bin/facilitator.rs:986]: You can, it doesn't require it.

File: facilitator/src/bin/facilitator.rs
Comment (aaomidi) [reply facilitator/src/bin/facilitator.rs:1597]: Ah yeah this was testing code

File: facilitator/src/kubernetes.rs
Comment (aaomidi) [reply facilitator/src/kubernetes.rs:79]: This is nicer!

File: manifest-updater/manifest/writer.go
Comment (aaomidi) [reply manifest-updater/manifest/writer.go:27]: I can change it to ioWriter

File: manifest-updater/manifest/writer.go
Comment (aaomidi) [reply manifest-updater/manifest/writer.go:47]: Do you envision the interface type returning a JSON string?

File: facilitator/taplo.toml
Comment (aaomidi) [reply facilitator/taplo.toml:1]: It's a small utility used for formatting toml files. I'll delete it here but we might want to introduce it later for our cargo file.

File: facilitator/src/bin/facilitator.rs
Comment (aaomidi) [reply facilitator/src/bin/facilitator.rs:915]: Correct!

File: facilitator/src/bin/facilitator.rs
Comment (aaomidi) [reply facilitator/src/bin/facilitator.rs:944]: I unwrap when ingestor_manifest_url is provided. If it's not then it goes to try to get the BSK from `batch_signing_key_from_arg`

Comment (tgeoghegan) [reply facilitator/src/bin/facilitator.rs:938]: Yeah but what if `namespace` is `None` and `own_manifest_url` is `Some`? Then we'll crash here. Admittedly `unwrap()` is used rather casually all over `facilitator.rs` so it's not the end of the world, but since this function returns `Result<BatchSigningKey>` already, it should be easy to return a proper error.

File: manifest-updater/manifest/writer.go
Comment (tgeoghegan) [reply manifest-updater/manifest/writer.go:45]: It seems like all you need `manifest` to be is something that can be passed to `json.Encoder.Encode()`. [`json.Encoder.Encode()`](https://golang.org/pkg/encoding/json/#Encoder.Encode) takes `interface{}`, so I think you could have a single method `func (w *Writer) WriteManifest(manifest interface{}, path string) error {}`. If you don't like `interface{}`, you could find (or define) some interface that both `DataShareProcessorSpecificManifest` and `IngestorGlobalManifest` implement so you can narrow this method's signature to implementations of that interface.

Comment (aaomidi) [reply manifest-updater/manifest/writer.go:26]: I still prefer being stricter with the types, so I took your advice but made it not-user-facing.

File: facilitator/src/bin/facilitator.rs
Comment (aaomidi) [reply facilitator/src/bin/facilitator.rs:938]: That's technically incorrect use of the CLI. I'll add an error message saying namespace must be supplied when using the `own_manifest_url` method.

I'm not super concerned about crashing immediately on startup since the recovery is nearly always going to be something the user did wrong when using the tool. I'll bubble the error up for the caller to handle.

File: facilitator/src/test_utils.rs
Comment (aaomidi) [reply facilitator/src/test_utils.rs:49]: It's a CSR, that name is uh, confusing.

File: terraform/modules/fake_server_resources/fake_server_resources.tf
Comment (aaomidi) [reply terraform/modules/fake_server_resources/fake_server_resources.tf:159]: I'll take a look into doing this!

File: manifest-updater/manifest/writer.go
Comment (tgeoghegan) [reply manifest-updater/manifest/writer.go:26]: Ooh, that's a neat compromise!

**Review by tgeoghegan — APPROVED** [review]

**Comment by aaomidi** [comment]
So before I merge this, my plan is to to make a testing environment using the latest code, and then deploy this on top. I want to make sure the deployment doesn't entirely break current environments. It's been a while since i've tried doing that.

**Comment by aaomidi** [comment]
This PR is way too big to rebase, I've honestly lost track of how many iterations it's been through. I'm going to squash and merge.

**Pull request merged**
